### PR TITLE
[WIP] Custom scoring

### PR DESCRIPTION
**Pull request by bartleyn**
## What does this PR do?
 Addresses #156 to allow for classification metrics from scikit-learn's metrics module to be passed to TPOT as scoring functions, along with keyword arguments for the function. Currently  lumps together scoring functions that require decision functions and prediction probabilities together. 

- [x] actual implementation 
- [x] unit-tests for different metrics
- [x] documentation change

## Where should the reviewer start?

fit(), _evaluate_individual(), and _train_model_and_predict()

## How should this PR be tested?

With classification scoring functions from sklearn.metrics

## Any background context you want to provide?

Some of the functions require a classifier's _decision-function_ output in place of a list of predicted classes, and some of the functions require a classifier's _prediction-probabilities_ output. These functions are currently lumped together in consideration -- it is difficult to inspect the input scoring function to see if it requires one or the other (without explicit listing of all the scoring functions, which would become bulky when we start considering regression metrics, etc), so scoring functions requiring decision functions _may_ mis-represent an individual's fitness.  

It was decided to lump decision function scoring functions into the prediction probability ones (and not vice versa), since decision functions can return negative numbers, and probabilities not. 

## What are the relevant issues?

#156 

## Questions:

- Do the docs need to be updated?
Yes, at the very least to reflect the caveats discussed above about the decision_function. 
- Does this PR add new (Python) dependencies?
No. 
- How much should TPOT handle wrt type checking, etc, and how much should we rely on the users?

### CODE REVIEW

**Comment by bartleyn** [comment]
After digging around for some code-inspection ideas, I found that you can access the docstring of the sklearn metrics with commands like `help()`. With some straightforward parsing, this would allow for flexibility when it comes to using a classifier's `decision_function` versus `predict_proba` (because the relevant sklearn metric docstrings specify which function to use), and eliminates the need to warn the user about this issue. I'll update accordingly.

**Comment by rhiever** [comment]
Hi @bartleyn, thank you so much for reworking the scoring system! As you may have seen, we've been refactoring TPOT so it's easier to maintain. You can find the latest version of the refactor in the [development branch](https://github.com/rhiever/tpot/tree/development).

This change will require some slight rework of your PR. Would you prefer to rework the PR into the new framework, or would you like @teaearlgraycold and I to take your code and rework it?

**Comment by bartleyn** [comment]
I've pulled those changes -- the refactor looks great. I should at least be able to resolve the merge conflicts in the main tpot.py (and maybe the tests). But maybe I'll hand it off at that point, since I'm not quite sure (yet) where things plug into the new framework.

**Comment by rhiever** [comment]
@teaearlgraycold can help with any complications arising from or explanations of the refactor. :-)

File: tpot/tpot.py [hunk 1]
```diff
@@ -485,314 +468,6 @@ def _parse_scoring_docstring(self, scoring_function):
...
-        """
-        learning_rate = min(1., max(learning_rate, 0.0001))
-        max_features = min(1., max(0., learning_rate))
-        min_weight = min(0.5, max(0., min_weight))
-
-        return self._train_model_and_predict(input_df, GradientBoostingClassifier,
-            learning_rate=learning_rate, n_estimators=500,
-            max_features=max_features, random_state=42, min_weight_fraction_leaf=min_weight)
-
     def _train_model_and_predict(self, input_df, model, **kwargs):
```
Comment (bartleyn) [inline tpot/tpot.py]: @teaearlgraycold  I resolved most conflicts (and tried to adopt the new numpy matrices/indices), but I left my original _train_model_and_predict() in because I wanted to see what you think the best way is for passing the clf_eval_func var to the separate classifier operators. In writing this I think I figured it out though -- throw it into a Classifier's _call and keep track of it there.

File: tpot/operators/classifiers/base.py [hunk 2]
```diff
@@ -57,8 +56,22 @@ def _train_and_predict(self, input_matrix, operator_args):
         clf.fit(self.training_features, self.training_classes)
 
         all_features = np.delete(input_matrix, non_feature_columns, axis=1)
-        input_matrix[:, GUESS_COL] = clf.predict(all_features)
+        if self.clf_eval_func == 'decision_function':
+            if 'decision_function' in dir(clf):
+                input_matrix = np.insert(input_matrix, GUESS_COL, clf.decision_function(all_features).T, axis=1)
```
Comment (bartleyn) [inline tpot/operators/classifiers/base.py]: @teaearlgraycold should we be updating the non_feature_columns in the classifiers that add multiple columns in at GUESS_COL?

File: tpot/tpot.py (hunk 1 above)
Comment (teaearlgraycold) [inline tpot/tpot.py]: Yeah that sounds good

File: tpot/operators/classifiers/base.py (hunk 2 above)
Comment (teaearlgraycold) [inline tpot/operators/classifiers/base.py]: >classifiers that add multiple columns in at `GUESS_COL`

I don't believe a classifier should be adding multiple columns in where `GUESS_COL` is.

Comment (bartleyn) [inline tpot/operators/classifiers/base.py]: Sorry about that. To clarify: certain scoring functions need an array of labels of size (n_samples,) as well as a matrix of prediction probabilities (or in some cases, decision function values) of size (n_samples, n_classes). This means using either the predict_proba() or decision_function() methods (and putting filler in if the classifier doesn't implement the function necessary). Originally I had been pickling these values into strings and putting them in the single guess column, but that was assuming an underlying dataframe. Rather than gutting the existing numpy matrix code, I figured it'd be easier to insert the (n_samples, n_classes) into the matrix at GUESS_COL and just try to keep track of all the added columns.

Comment (teaearlgraycold) [inline tpot/operators/classifiers/base.py]: If you need to store more data you'll probably be best off adding another "non_feature_column" entirely - not just keeping a reference to a column after `GUESS_COL`. I never really considered that contingency, though, so that might break all other code dealing with column indices.

Comment (rhiever) [inline tpot/operators/classifiers/base.py]: For now, I think we should only support outputs and scorers that output only a single value. Multi-label classification may eventually require a set up that supports multiple outputs, but that's a fair bit down the line.

Comment (bartleyn) [inline tpot/operators/classifiers/base.py]: You all helped me realize that I was totally overthinking it/ignoring the bigger picture. I think a solution to make this work is to still insert the classifier's `predict()` outputs in GUESS_COL, but to tack on anything regarding `predict_proba()` or `decision_function()` to the end of the matrix, and to use that slice at the end in the scoring functions as necessary. I've updated accordingly.

**Comment by rhiever** [comment]
@bartleyn: Just merged this PR and @teaearlgraycold will jump in to wrap it up for the 0.5 release (hopefully this week). Thank you again!
